File: AlgorithmLcr/Try2-Known component types, quantities and parameters/netgraph_id/enumerate.py
"""Isomorph-free two-phase enumeration of 2-terminal networks (DESIGN.md 4).

Phase 1 (structure layer): for V = 2..E+1 enumerate every multiset of E
edges over the S(V) = V(V-1)/2 node-pair slots (combinations with
replacement), keep connected R0-clean structures, canonicalize (minimum
over the relabeling group) and deduplicate.  Every isomorphism class of
admissible uniform structures is emitted exactly once (completeness proof
in DESIGN.md 4.2).

Phase 2 (assignment layer): for each structure, assign the concrete known
component multiset to the edge instances.  Components are applied in the
canonical sorted order; equal components are interchangeable; permutations
are deduplicated by the canonical per-slot serialization minimized over the
structure automorphism group.
"""

from __future__ import annotations

from functools import lru_cache
from itertools import combinations_with_replacement, permutations

from .components import ComponentSet
from .graph import (Structure, Network, canonical_mult, empty_mult, is_connected,
                    has_dead_part, make_structure, n_slots, permute_slot_keys,
                    slot_list)


@lru_cache(maxsize=64)
def enumerate_structures(E: int, *, allow_dead: bool = False) -> tuple[Structure, ...]:
    """All admissible 2-terminal multigraph structures with exactly E edges.

    Result is cached per (E, allow_dead); sorted by (V, mult) for
    determinism.
    """
    if E < 1:
        raise ValueError("E must be >= 1")
    out: list[Structure] = []
    for V in range(2, E + 2):
        S = n_slots(V)
        seen: set[tuple[int, ...]] = set()
        for combo in combinations_with_replacement(range(S), E):
            mult = empty_mult(V)
            for k in combo:
                mult[k] += 1
            if not is_connected(V, mult):
                continue
            if not allow_dead and has_dead_part(V, mult):
                continue
            cm = canonical_mult(V, mult)
            if cm in seen:
                continue
            seen.add(cm)
            out.append(make_structure(V, cm, canonicalize=False))
    out.sort(key=lambda st: (st.V, st.mult))
    return tuple(out)


def iter_assignments(structure: Structure, compset: ComponentSet):
    """Yield all distinct component assignments (tuples of component indices
    per edge instance) for one structure, one per wiring-equivalence orbit.

    Dedup keys: per-slot sorted component-key tuples over the FULL slot
    layout (empty slots are ()), minimized over Aut when |Aut| > 1.
    Parallel-edge orderings collapse via the within-slot sort.
    """
    comps = compset.components
    E = structure.n_edges
    V = structure.V
    if E != len(comps):
        raise ValueError("structure edge count != component count")
    keys = [c.key() for c in comps]

    inst_by_slot: list[list[int]] = [[] for _ in range(n_slots(V))]
    for t, slot in enumerate(structure.slot_of_instances()):
        inst_by_slot[slot].append(t)

    aut = structure.aut if len(structure.aut) > 1 else ()
    seen: set = set()
    for perm in permutations(range(E)):
        ser = tuple(tuple(sorted(keys[perm[t]] for t in occ)) if occ else ()
                    for occ in inst_by_slot)
        if aut:
            ser = min(permute_slot_keys(V, ser, p) for p in aut)
        if ser in seen:
            continue
        seen.add(ser)
        yield perm


def count_assignments(structure: Structure, compset: ComponentSet) -> int:
    """Number of distinct assignments (consumes the iterator)."""
    return sum(1 for _ in iter_assignments(structure, compset))


def iter_networks(compset: ComponentSet, structures) :
    """Stream all candidate Networks for a component set."""
    for st in structures:
        for assign in iter_assignments(st, compset):
            yield Network(structure=st, assign=assign)

File: AlgorithmLcr/Try2-Known component types, quantities and parameters/cppversion/src/enumerate.cpp
#include "enumerate.hpp"

#include <algorithm>
#include <map>
#include <mutex>
#include <numeric>
#include <set>
#include <stdexcept>

namespace ng {

namespace {

// All multisets of E indices from [0, S) in lexicographic (non-decreasing)
// order -- the combinations_with_replacement order.
void combosWithReplacement(int S, int E, const std::function<void(const std::vector<int>&)>& yield) {
    std::vector<int> combo((size_t)E, 0);
    yield(combo);
    if (E == 0) return;
    while (true) {
        // find rightmost element that can be incremented
        int i = E - 1;
        while (i >= 0 && combo[i] == S - 1) --i;
        if (i < 0) return;
        int v = combo[i] + 1;
        for (int k = i; k < E; ++k) combo[k] = v;
        yield(combo);
    }
}

std::vector<Structure> computeStructures(int E, bool allowDead) {
    if (E < 1) throw std::invalid_argument("E must be >= 1");
    std::vector<Structure> out;
    for (int V = 2; V <= E + 1; ++V) {
        const int S = nSlots(V);
        std::set<std::vector<int>> seen;
        combosWithReplacement(S, E, [&](const std::vector<int>& combo) {
            std::vector<int> mult = emptyMult(V);
            for (int k : combo) mult[k] += 1;
            if (!isConnected(V, mult)) return;
            if (!allowDead && hasDeadPart(V, mult)) return;
            std::vector<int> cm = canonicalMult(V, mult);
            if (seen.count(cm)) return;
            seen.insert(cm);
            out.push_back(makeStructure(V, cm, false));
        });
    }
    std::sort(out.begin(), out.end(), [](const Structure& a, const Structure& b) {
        if (a.V != b.V) return a.V < b.V;
        return a.mult < b.mult;
    });
    return out;
}

}  // namespace

const std::vector<Structure>& enumerateStructures(int E, bool allowDead) {
    static std::mutex mu;
    static std::map<std::pair<int, bool>, std::vector<Structure>> cache;
    std::lock_guard<std::mutex> lock(mu);
    auto key = std::make_pair(E, allowDead);
    auto it = cache.find(key);
    if (it == cache.end())
        it = cache.emplace(key, computeStructures(E, allowDead)).first;
    return it->second;
}

void iterAssignments(const Structure& structure, const ComponentSet& compset,
                     const std::function<void(const std::vector<int>&)>&& yield) {
    const auto& comps = compset.components();
    const int E = structure.nEdges();
    const int V = structure.V;
    if (E != compset.n())
        throw std::invalid_argument("structure edge count != component count");
    if (E == 0) return;

    const auto soi = structure.slotOfInstances();
    std::vector<std::vector<int>> instBySlot((size_t)nSlots(V));
    for (int t = 0; t < E; ++t) instBySlot[soi[t]].push_back(t);

    const bool useAut = structure.aut.size() > 1;
    std::set<std::vector<std::vector<Component>>> seen;

    std::vector<int> perm((size_t)E);
    std::iota(perm.begin(), perm.end(), 0);  // itertools.permutations order
    do {
        std::vector<std::vector<Component>> ser((size_t)nSlots(V));
        for (int slot = 0; slot < nSlots(V); ++slot) {
            const auto& occ = instBySlot[slot];
            if (occ.empty()) continue;
            auto& g = ser[slot];
            g.reserve(occ.size());
            for (int t : occ) g.push_back(comps[perm[t]]);
            std::sort(g.begin(), g.end());
        }
        if (useAut) {
            std::vector<std::vector<Component>> best = ser;
            for (const auto& p : structure.aut) {
                auto m = permuteSlotKeys(V, ser, p);
                if (m < best) best = std::move(m);
            }
            ser = std::move(best);
        }
        if (seen.count(ser)) continue;
        seen.insert(std::move(ser));
        yield(perm);
    } while (std::next_permutation(perm.begin(), perm.end()));
}

int countAssignments(const Structure& structure, const ComponentSet& compset) {
    int n = 0;
    iterAssignments(structure, compset, [&](const std::vector<int>&) { ++n; });
    return n;
}

}  // namespace ng
